ACPs/108-evm-event-importing/README.md

```diff
@@ -0,0 +1,117 @@
+```text
+ACP: 108
+Title: EVM Event Importing Standard
+Author(s): Michael Kaplan (https://github.com/mkaplan13)
+Discussions-To: https://github.com/avalanche-foundation/ACPs/discussions/109
+Status: Proposed
+Track: Best Practices Track
+```
+
+## Abstract
+
+Defines a standard smart contract interface and abstract implementation for importing EVM events from any blockchain within Avalanche using [Avalanche Warp Messaging](https://docs.avax.network/build/cross-chain/awm/overview).
+
+## Motivation
+
+The implementation of Avalanche Warp Messaging within `coreth` and `subnet-evm` exposes a [mechanism for getting authenticated hashes of blocks](https://github.com/ava-labs/subnet-evm/blob/master/contracts/contracts/interfaces/IWarpMessenger.sol#L43) that have been accepted on blockchains within Avalanche. There is currently no clear standard for using authenticated block hashes in smart contracts within Avalanche, making it difficult to build applications that leverage this mechanism. In order to make effective use of authenticated block hashes, contracts must be provided encoded block headers that match the authenticated block hashes and also Merkle proofs that are verified against the state or receipts root contained in the block header. 
+
+With a standard interface and abstract contract implemetation that handles the authentication of block hashes and verification of Merkle proofs, smart contract developers on Avalanche will be able to much more easily create applications that leverage data from other Avalanche blockchains. These type of cross-chain application do not require any direct interaction on the source chain.
+
+## Specification
+
+### Event Importing Interface
+
+We propose that smart contracts importing EVM events emitted by other blockchains within Avalanche implement the following interface.
+
+#### Methods
+
+Imports the EVM event uniquely identified by the source blockchain ID, block header, transaction index, and log index. The event data is provided in the `receiptProof`, which must be a valid Merkle proof of the receipt for the given `txIndex` against the receipt root of the `blockHeader`. Must emit an `EventImported` event upon success. Should validate that the `blockHeader` matches an authenticated block hash from the `sourceBlockchainID`.
+```solidity
+function importEvent(
+    bytes32 sourceBlockchainID,
+    bytes calldata blockHeader,
+    uint256 txIndex,
+    bytes[] calldata receiptProof,
+    uint256 logIndex
+) external;
+```
+
+This interface does not require that the Warp precompile is used to authenticate block hashes. Implementations could:
+- Use an external block hash verification mechanism, such as trusted oracles.
+- Use the Warp precompile to authenticate block hashes provided in the transaction calling `importEvent`.
+- Look up previously authenticated block hashes from another contract, such that multiple events could be imported from the same block hash, which would only need to be authenticated once.
```

Comment:
Think it may be best to move this topic to a discussions thread once this PR is merged, but:

Looking at the `importEvents` interface:
```
function importEvent(
    bytes32 sourceBlockchainID,
    bytes calldata blockHeader,
    uint256 txIndex,
    bytes[] calldata receiptProof,
    uint256 logIndex
) external;
```

The pieces that seem EVM specific are the `txIndex` and `logIndex`. The `txIndex` is actually the key of trie that the `receiptProof` proves inclusion of, so maybe that should be renamed `trieKey` (and possibly made `bytes`?). The `logIndex` is extra EVM specific information for knowing how to interpret the data imported using the proof, so that could also be made more generalized. 

I still need to think on it more, but right now I think a reasonable set of assumptions made by the interface would be:
- The hash imported from the `sourceBlockchainID` is the hash of an encoded struct, and the `blockHeader` is the pre-image of that hash.
- The `blockHeader` includes a root hash of a trie.
- The `txIndex` is the key of that trie that `receiptProof` proves the inclusion of.

The tweaks to the interface to reduce it that set of assumptions would probably be:
- Renaming a few params
- Making `txIndex` bytes
- Moving `logIndex` to an extra data field (?)

Curious what your thoughts here are.

Reply:
This makes sense to me.

It may be that if we try to support arbitrary VMs, this just turns into a router that can call on a specific implementation or verifier. I'm not sure the resulting data (the receipts in this case), would look the same for every VM. If the inputs and outputs both look different, I'm not sure yet if putting a router in the middle will make sense or whether we should just have different implementations that do not follow the same interface as needed.